File: lib/screens/redesign/redesign_bridge.dart
// Live-data bridge for the SG Transit redesign (Flutter).
//
// Mirrors ios-native/Leyne/Redesign/RedesignBridge.swift: adapts the app's
// real LTA DataMall domain types (Service / NearbyStop / MrtStation) into the
// redesign's compact Rd* view-models, so the redesign renders genuine live data
// without rewriting every screen. No mock content here — everything flows from
// DataStore.shared.

import 'package:flutter/material.dart';

import '../../data/models.dart';
import '../../data/mrt_stations.dart';
import 'redesign_data.dart';
import 'redesign_theme.dart';

RdLoad rdLoad(Load l) => switch (l) {
      Load.sea => RdLoad.seats,
      Load.sda => RdLoad.standing,
      Load.lsd => RdLoad.packed,
    };

/// Whole-minute ETA label; "0" means arriving now.
String rdMinLabel(int etaSec) => '${((etaSec + 30) ~/ 60).clamp(0, 999)}';

RdArrival rdArrival(Service s) => RdArrival(
      route: s.no,
      dest: s.dest,
      load: rdLoad(s.load),
      min: rdMinLabel(s.etaSec),
      then: s.followingSec > 0 ? 'then ${((s.followingSec + 30) ~/ 60).clamp(1, 999)}' : null,
    );

RdStop rdStop(NearbyStop n) {
  final here = n.distanceM <= 40;
  return RdStop(
    name: n.stopName,
    code: n.stopCode,
    dist: here ? "You're at this stop" : '${n.walkMin} min walk · ${fmtDistance(n.distanceM)}',
    distShort: here ? 'You are here' : fmtDistance(n.distanceM),
    badge: here ? "YOU'RE HERE" : '${n.walkMin} MIN WALK',
    arrivals: n.services.map(rdArrival).toList(growable: false),
  );
}

/// MRT line code(s) for an interchange bus-stop description, or [] otherwise.
List<MrtCode> rdMrtBadges(String stopName) => resolveMrtStation(stopName)?.codes ?? const [];

/// Readable foreground for a line-coloured badge — dark ink on the light
/// (orange) Circle Line, white on everything else.
Color rdMrtBadgeFg(String code) {
  final prefix = (code.length >= 2 ? code.substring(0, 2) : code).toUpperCase();
  return (prefix == 'CC' || prefix == 'CE') ? const Color(0xFF38240A) : Colors.white;
}

/// Small line-coloured code chip(s) shown beside an interchange stop name
/// (item 3). Renders nothing when the stop isn't a recognised rail station.
class RdMrtBadgeRow extends StatelessWidget {
  const RdMrtBadgeRow({super.key, required this.stopName, this.size = 10});

  final String stopName;
  final double size;

  @override
  Widget build(BuildContext context) {
    final codes = rdMrtBadges(stopName);
    if (codes.isEmpty) return const SizedBox.shrink();
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        for (final c in codes)
          Padding(
            padding: const EdgeInsets.only(left: 4),
            child: Container(
              padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
              decoration: BoxDecoration(color: c.color, borderRadius: BorderRadius.circular(6)),
              child: Text(c.code,
                  style: rdText(size: size, weight: FontWeight.w800, color: rdMrtBadgeFg(c.code))),
            ),
          ),
      ],
    );
  }
}

File: ios-native/Leyne/Redesign/RedesignBridge.swift
// Live-data bridge for the SG Transit redesign (iOS).
//
// The redesign's view layer is written against the compact `RD*` view-models
// (RDStop / RDArrival / RDStation …). This file adapts the app's real LTA
// DataMall domain types (Service / NearbyStop / MrtGeoStation / TrainAlert)
// into those view-models, so the redesign renders genuine live data without
// rewriting every screen. There is no mock content here — everything flows
// from `DataStore.shared`.

import SwiftUI
import CoreLocation

// MARK: - Occupancy

func rdLoad(_ l: Load) -> RDLoad {
    switch l {
    case .sea: return .seats
    case .sda: return .standing
    case .lsd: return .packed
    }
}

/// Whole-minute ETA label for an arrival in seconds. "0" means arriving now.
func rdMinLabel(_ etaSec: Int) -> String {
    String(max(0, (etaSec + 30) / 60))
}

// MARK: - Arrivals

func rdArrival(_ s: Service) -> RDArrival {
    let then = s.followingSec > 0 ? "then \(max(1, (s.followingSec + 30) / 60))" : nil
    return RDArrival(route: s.no, dest: s.dest, load: rdLoad(s.load),
                     min: rdMinLabel(s.etaSec), then: then)
}

func rdArrivals(_ services: [Service]) -> [RDArrival] { services.map(rdArrival) }

// MARK: - Stops

/// Map a live nearby stop into the redesign's RDStop view-model.
func rdStop(_ n: NearbyStop) -> RDStop {
    let here = n.distanceM <= 40
    return RDStop(
        name: n.stopName,
        code: n.stopCode,
        dist: here ? "You're at this stop" : "\(n.walkMin) min walk · \(fmtDistance(n.distanceM))",
        distShort: here ? "You are here" : fmtDistance(n.distanceM),
        badge: here ? "YOU'RE HERE" : "\(n.walkMin) MIN WALK",
        arrivals: rdArrivals(n.services))
}

// MARK: - MRT interchange badges (item 3)

/// The rail line code(s) + colours for a bus-stop description, or [] when the
/// stop is not (next to) a recognised MRT/LRT station. Drives the small
/// line-coloured chip shown beside interchange stop names.
func rdMrtBadges(forStopNamed name: String) -> [MrtCode] {
    resolveMrtStation(name)?.codes ?? []
}

/// Readable foreground for a line-coloured badge — dark ink on the light
/// (orange) Circle Line, white on everything else.
func rdMrtBadgeFg(_ code: String) -> Color {
    let prefix = String(code.prefix(2)).uppercased()
    return (prefix == "CC" || prefix == "CE")
        ? Color(.sRGB, red: 0.22, green: 0.14, blue: 0, opacity: 1)
        : .white
}

/// Small line-coloured code chip(s) shown beside an interchange stop name
/// (item 3). Renders nothing when the stop isn't a recognised rail station.
struct RDMrtBadgeRow: View {
    let stopName: String
    var size: CGFloat = 10

    var body: some View {
        let codes = rdMrtBadges(forStopNamed: stopName)
        HStack(spacing: 4) {
            ForEach(codes, id: \.code) { c in
                Text(c.code)
                    .font(.system(size: size, weight: .heavy))
                    .foregroundStyle(rdMrtBadgeFg(c.code))
                    .padding(.horizontal, 6).padding(.vertical, 2)
                    .background(c.color)
                    .clipShape(RoundedRectangle(cornerRadius: 6, style: .continuous))
            }
        }
    }
}
